### ISSUE TITLE
Is --invert_hash not compatible for multi pass models on Linear Regression?

### ISSUE THREAD

**Issue by kfoofw** [VowpalWabbit/vowpal_wabbit#2216]
Hi there!

I am trying to apply vowpalwabbit as a linear regression model (with the Boston housing toy dataset) and I came across this SO post reply (https://stackoverflow.com/questions/19187532/ordinary-least-squares-regression-in-vowpal-wabbit). Note that this SO post answer was in 2014, when the published version of vowpal was 7.3.1 while my current version is 8.5.0. 

As part of reproducibility, the command line code that I used is similar to what was suggested in the post:
```
vw -d train.txt -f linear_model -c --passes 50 --holdout_off --loss_function squared --invert_hash model_readable.txt
```
However, when I input the arguments for multi pass training with --invert_hash, it gave me an error:

> Error: invert_hash is incompatible with a cache file.  Use it in single pass mode only.

Alternatively, I also came across another SO post reply(https://stackoverflow.com/questions/24437152/vowpal-wabbit-inverted-hash-option-produces-empty-output-but-why) on Logistic Regression multi-pass by Martin who suggested creating the binary model file before converting to inverted hash. This post was also posted in 2014. 

When I used the following code (as suggested for creating the binary model of multi-pass before converting to inverted_hash on one-pass testing): 
```
vw -d train.txt -c --passes 10 -f model.binary
vw -d train.txt -t -i model.binary --invert_hash linear_model.humanreadable
```
This resulted in a model output that purely predicted 4.74 for the testing of every observation of the training data.

> only testing
> Num weight bits = 18
> learning rate = 0.5
> initial_t = 0
> power_t = 0.5
> using no cache
> Reading datafile = train.txt
> num sources = 1
> average  since         example        example  current  current  current
> loss     last          counter         weight    label  predict features
> 214.916992 214.916992            1            1.0  19.4000   4.7400       12
> 145.827717 76.738441            2            2.0  13.5000   4.7400       12
> 640.982603 1136.137489            4            4.0  19.7000   4.7400       12
> 459.218392 277.454182            8            8.0  23.1000   4.7400       12
> 521.886717 584.555042           16           16.0   8.7000   4.7400       12
> 472.089673 422.292630           32           32.0  18.3000   4.7400       12
> 452.764611 433.439549           64           64.0  27.5000   4.7400       12
> 388.353522 323.942433          128          128.0  18.1000   4.7400       12
> 387.781603 387.209684          256          256.0  20.8000   4.7400       12

The output of "linear_model.humanreadable" is:

> Version 8.5.0
> Id 
> Min label:0
> Max label:50
> bits:18
> lda:0
> 0 ngram:
> 0 skip:
> options:
> Checksum: 2309139350
> :0
> Constant:116060:3.52267
> 

I suspected that this was used some form of overfitting, so instead I adjusted the multi-pass to 1:
```
vw -d train.txt -c --passes 1 -f model.binary
vw -d train.txt -t -i model.binary --invert_hash linear_model.humanreadable
```
The trained single pass model still seemed to be predicting a constant value for testing of every training data.

> only testing
> Num weight bits = 18
> learning rate = 0.5
> initial_t = 0
> power_t = 0.5
> using no cache
> Reading datafile = train.txt
> num sources = 1
> average  since         example        example  current  current  current
> loss     last          counter         weight    label  predict features
> 252.089630 252.089630            1            1.0  19.4000   3.5227       12
> 175.818382 99.547134            2            2.0  13.5000   3.5227       12
> 693.371349 1210.924316            4            4.0  19.7000   3.5227       12
> 506.251062 319.130775            8            8.0  23.1000   3.5227       12
> 570.836604 635.422145           16           16.0   8.7000   3.5227       12
> 519.730976 468.625349           32           32.0  18.3000   3.5227       12
> 500.398302 481.065627           64           64.0  27.5000   3.5227       12
> 431.747640 363.096978          128          128.0  18.1000   3.5227       12
> 431.430.951856          256          256.0  20.8000   3.5227       12

The updated output of "linear_model.humanreadble" still reads similarly:

> Version 8.5.0
> Id 
> Min label:0
> Max label:50
> bits:18
> lda:0
> 0 ngram:
> 0 skip:
> options:
> Checksum: 2309139350
> :0
> Constant:116060:3.52267

This "single" multipass model seems different from the direct simple invert_hash model as shown:
```
vw -d train.txt --loss_function squared --invert_hash linear_model_IH 
```
where the outputs of the "linear_model_IH" reads:

> Version 8.5.0
> Id 
> Min label:0
> Max label:50
> bits:18
> lda:0
> 0 ngram:
> 0 skip:
> options:
> Checksum: 2309139350
> :0
> Constant:116060:3.51799
> Num_feat^AGE:191433:0.0242481
> Num_feat^B:82735:0.00921203
> Num_feat^CHAS:63356:0.874292
> Num_feat^CRIM:227021:-0.00192302
> Num_feat^DIS:61172:0.381227
> Num_feat^INDUS:218500:0.0596201
> Num_feat^LSTAT:52422:0.0507423
> Num_feat^NOX:209497:3.43052
> Num_feat^PTRATIO:246628:0.158207
> Num_feat^RAD:207988:0.0610307
> Num_feat^RM:85721:0.397752
> Num_feat^TAX:114825:0.0034895
> Num_feat^ZN:46011:0.0256615
> 

Is there a proper option for creating inverted hash outputs for multi-pass models? I suspect there might be an issue with my command line inputs, so I appreciate any corrections on my code. 

Thanks in advance!

**Comment by jackgerrits** [comment]
I'm not too sure about the issue with the consistent 4.74 predictions - @marco-rossi29, any ideas there?

That behavior should be working in current master (with one small caveat I'll address below), so perhaps there was an issue in 8.5.0. Can you try from master?

However, for the invert_hash if there is no known hash for a feature when you do invert_hash current behavior (master) is to skip it. I have a PR out to fix this, so if there is no hash it will still show the weight but without the inverted hash name. See #2246.

**Comment by kfoofw** [comment]
thanks @jackgerrits. I'll update my vowpal wabbit version to the current master and update any findings on this!

**Comment by kfoofw** [comment]
Hi @jackgerrits, thanks for your help. The latest version of vowpalwabbit on the repo master (8.8.0) solved the issue of the constant prediction. 

As part of the update, here's a description of the issues that I encountered in 8.5.0, which were fixed using vw=8.8.0.
- Constant prediction issue using multi pass model is fixed
- Invert hash on multipass of 1 pass binary model versus implicit single pass model with invert hash now shows the same feature weights.

________________________________________________

I ran a multipass training with 1 single pass and created a model binary file before using invert-hash on the model binary file to check out the "inverted" feature weights.

`vw -d train.txt -c --passes 1 -f model.binary`
`vw -d train.txt -t -i model.binary --invert_hash linear_model.humanreadable`

> only testing
> Num weight bits = 18
> learning rate = 0.5
> initial_t = 0
> power_t = 0.5
> using no cache
> Reading datafile = train.txt
> num sources = 1
> average  since         example        example  current  current  current
> loss     last          counter         weight    label  predict features
> 51.742210 51.742210            1            1.0  14.3000  21.4932       12
> 41.833765 31.925320            2            2.0  23.7000  18.0498       12
> 24.641104 7.448442            4            4.0  18.9000  21.1955       13
> 22.347419 20.053735            8            8.0  13.4000  22.0622       12
> 31.149827 39.952235           16           16.0  21.7000  18.6151       12
> 36.570957 41.992087           32           32.0  23.8000  20.7387       12
> 83.146581 129.722204           64           64.0  12.0000  23.4474       12
> 117.810741 152.474901          128          128.0  33.3000  20.5005       13
> 108.177625 98.544508          256          256.0  16.1000  18.0207       12
> 
> finished run
> number of examples = 404
> weighted example sum = 404.000000
> weighted label sum = 9168.200138
> average loss = 97.738202
> best constant = 22.693565
> total feature number = 4981

As shown, the "test only" prediction using the binary model no longer has a fixed constant prediction that was previously encountered in 8.5.0.

The invert_hash of the binary model file is stored in "linear_model.human_readable", which shows the following:

> Version 8.8.0
> Id 
> Min label:0
> Max label:50
> bits:18
> lda:0
> 0 ngram:
> 0 skip:
> options:
> Checksum: 2309139350
> :0
> Constant:116060:3.55963
> Num_feat^AGE:191433:0.0274775
> Num_feat^B:82735:0.00969477
> Num_feat^CHAS:63356:1.40884
> Num_feat^CRIM:227021:-0.00805428
> Num_feat^DIS:61172:0.300686
> Num_feat^INDUS:218500:0.0657379
> Num_feat^LSTAT:52422:0.0302051
> Num_feat^NOX:209497:3.65656
> Num_feat^PTRATIO:246628:0.153114
> Num_feat^RAD:207988:0.0559221
> Num_feat^RM:85721:0.424642
> Num_feat^TAX:114825:0.00366466
> Num_feat^ZN:46011:0.0239644

I corroborated this by running a direct invert hash on a implicit single pass model with the following commands:

`vw -d train.txt --loss_function squared --invert_hash linear_model_IH `

The output of the invert_hash is stored in the file "linear_model_IH" which reads the following:

> Version 8.8.0
> Id 
> Min label:0
> Max label:50
> bits:18
> lda:0
> 0 ngram:
> 0 skip:
> options:
> Checksum: 2309139350
> :0
> Constant:116060:3.55963
> Num_feat^AGE:191433:0.0274775
> Num_feat^B:82735:0.00969477
> Num_feat^CHAS:63356:1.40884
> Num_feat^CRIM:227021:-0.00805428
> Num_feat^DIS:61172:0.300686
> Num_feat^INDUS:218500:0.0657379
> Num_feat^LSTAT:52422:0.0302051
> Num_feat^NOX:209497:3.65656
> Num_feat^PTRATIO:246628:0.153114
> Num_feat^RAD:207988:0.0559221
> Num_feat^RM:85721:0.424642
> Num_feat^TAX:114825:0.00366466
> Num_feat^ZN:46011:0.0239644

This is similar to the invert_hash output of the multipass of 1 pass model binary file.
____________________________________________________________

Thanks @jackgerrits once again for your help!
